swap/simulations_test.go

```diff
@@ -676,33 +683,60 @@ func TestBasicSwapSimulation(t *testing.T) {
 	log.Info("Simulation ended")
 }
 
-func waitForChequeProcessed(t *testing.T, ts *testService, counter metrics.Counter, lastCount int64) error {
+func waitForChequeProcessed(t *testing.T, backend *swapTestBackend, counter metrics.Counter, lastCount int64, p *Peer, expectedLastPayout uint64) error {
 	t.Helper()
 	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
 	defer cancel()
 
 	// we are going to wait for two things:
-	// * that a cheque has been processed
+	// * that a cheque has been processed and confirmed
 	// * that the other side actually did process the testMsgPrice message
 	//   (by checking that the message counter has been increased)
 	var wg sync.WaitGroup
-	wg.Add(2)
+	errs := []string{}
+	wg.Add(3)
 
 	go func() {
-		backend := ts.swap.backend.(*swapTestBackend)
-		select {
-		case <-ctx.Done():
-			t.Fatal("timed out waiting for cheque to be processed")
-		case <-backend.cashDone:
-			wg.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				errs = append(errs, "Timed out waiting for cheque to have been cached")
+				wg.Done()
+				return
+			case <-backend.cashDone:
+				wg.Done()
+				return
+			}
+		}
+	}()
+
+	go func() {
```

Comment:
This goroutine creates a data race. It can be solved by setting lock on Peer around getLastSentCumulativePayout a few lines below.

```
ARNING: DATA RACE
Write at 0x00c0120eefd8 by goroutine 144:
  github.com/ethersphere/swarm/swap.(*Peer).setLastSentCheque()
      /Users/janos/go/src/github.com/ethersphere/swarm/swap/peer.go:105 +0x5e
  github.com/ethersphere/swarm/swap.(*Swap).handleConfirmChequeMsg()
      /Users/janos/go/src/github.com/ethersphere/swarm/swap/swap.go:424 +0x171

Previous read at 0x00c0120eefd8 by goroutine 142:
  github.com/ethersphere/swarm/swap.waitForChequeProcessed.func2()
      /Users/janos/go/src/github.com/ethersphere/swarm/swap/peer.go:86 +0x87

Goroutine 144 (running) created at:
  github.com/ethersphere/swarm/swap.(*Swap).handleMsg.func1()
      /Users/janos/go/src/github.com/ethersphere/swarm/swap/swap.go:335 +0xdc
  github.com/ethersphere/swarm/p2p/protocols.(*Peer).handleIncoming()
      /Users/janos/go/src/github.com/ethersphere/swarm/p2p/protocols/protocol.go:338 +0xc2d
  github.com/ethersphere/swarm/p2p/protocols.(*Peer).Run()
      /Users/janos/go/src/github.com/ethersphere/swarm/p2p/protocols/protocol.go:235 +0x4c
  github.com/ethersphere/swarm/swap.(*Swap).run()
      /Users/janos/go/src/github.com/ethersphere/swarm/swap/protocol.go:131 +0x5bb
  github.com/ethersphere/swarm/swap.(*Swap).run-fm()
      /Users/janos/go/src/github.com/ethersphere/swarm/swap/protocol.go:105 +0x63
  github.com/ethereum/go-ethereum/p2p.(*Peer).startProtocols.func1()
      /Users/janos/go/pkg/mod/github.com/ethereum/go-ethereum@v1.9.2/p2p/peer.go:363 +0x8d

Goroutine 142 (running) created at:
  github.com/ethersphere/swarm/swap.waitForChequeProcessed()
      /Users/janos/go/src/github.com/ethersphere/swarm/swap/simulations_test.go:713 +0x23c
  github.com/ethersphere/swarm/swap.TestMultiChequeSimulation()
      /Users/janos/go/src/github.com/ethersphere/swarm/swap/simulations_test.go:460 +0xb2a
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x192
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x284
  fmt.(*ss).doScanf()
      /usr/local/go/src/fmt/scan.go:1217 +0x382
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x192
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  github.com/syndtr/goleveldb/leveldb.(*DB).checkAndCleanFiles()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db_util.go:52 +0x2f8
  github.com/syndtr/goleveldb/leveldb.openDB()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db.go:130 +0x9f0
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x192
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x192
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x284
  fmt.(*ss).doScanf()
      /usr/local/go/src/fmt/scan.go:1217 +0x382
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x192
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  github.com/syndtr/goleveldb/leveldb.(*DB).recoverJournal()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db.go:476 +0xc5
  github.com/syndtr/goleveldb/leveldb.openDB()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db.go:125 +0x9cc
  github.com/syndtr/goleveldb/leveldb.Open()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db.go:197 +0x1d3
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x192
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  github.com/syndtr/goleveldb/leveldb.(*session).recover.func1()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/session.go:134 +0xf4
  github.com/syndtr/goleveldb/leveldb.(*session).recover()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/session.go:142 +0x1938
  github.com/syndtr/goleveldb/leveldb.Open()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db.go:183 +0xf8
  github.com/syndtr/goleveldb/leveldb.OpenFile()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db.go:219 +0xa5
  github.com/ethersphere/swarm/swap.newSharedBackendSwaps()
      /Users/janos/go/src/github.com/ethersphere/swarm/state/dbstore.go:49 +0x723
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x192
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x284
  fmt.(*ss).doScanf()
      /usr/local/go/src/fmt/scan.go:1217 +0x382
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x192
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  github.com/syndtr/goleveldb/leveldb.(*DB).checkAndCleanFiles()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db_util.go:52 +0x2f8
  github.com/syndtr/goleveldb/leveldb.openDB()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db.go:130 +0x9f0
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x192
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x192
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x284
  fmt.(*ss).doScanf()
      /usr/local/go/src/fmt/scan.go:1217 +0x382
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x192
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  github.com/syndtr/goleveldb/leveldb.(*DB).recoverJournal()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db.go:476 +0xc5
  github.com/syndtr/goleveldb/leveldb.openDB()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db.go:125 +0x9cc
  github.com/syndtr/goleveldb/leveldb.Open()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db.go:197 +0x1d3
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xcf
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /usr/local/go/src/fmt/scan.go:114 +0x192
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x2c4
  github.com/syndtr/goleveldb/leveldb.(*session).recover.func1()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/session.go:134 +0xf4
  github.com/syndtr/goleveldb/leveldb.(*session).recover()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/session.go:142 +0x1938
  github.com/syndtr/goleveldb/leveldb.Open()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db.go:183 +0xf8
  github.com/syndtr/goleveldb/leveldb.OpenFile()
      /Users/janos/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20190318030020-c3a204f8e965/leveldb/db.go:219 +0xa5
  github.com/ethersphere/swarm/swap.newSharedBackendSwaps()
      /Users/janos/go/src/github.com/ethersphere/swarm/state/dbstore.go:49 +0x723
  github.com/ethersphere/swarm/swap.TestMultiChequeSimulation()
      /Users/janos/go/src/github.com/ethersphere/swarm/swap/simulations_test.go:376 +0x5c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199
==================
```